File: 2025/JuyoungMoon/week4/2025-06-29-guess-number.cpp
// Author: Juyoung Moon

// KDD LeetCode Study Week 4: Binary Search.
// https://github.com/juyomo/leetcode-study

// LeetCode #374.
// https://leetcode.com/problems/guess-number-higher-or-lower/

/** 
 * Forward declaration of guess API.
 * @param  num   your guess
 * @return 	     -1 if num is higher than the picked number
 *			      1 if num is lower than the picked number
 *               otherwise return 0
 * int guess(int num);
 */

class Solution {
public:
    int guessNumber(int n) {
        return guessHelper(1, n);
    }

    int guessHelper(int low, int high) {
        int mid = low + (high - low) / 2;
        int res = guess(mid);
        if (res == 0) {
            return mid;
        } else if (res == 1) {
            return guessHelper(mid + 1, high);
        } else {
            return guessHelper(low, mid - 1);
        }
    }
};

File: 2025/SangilPark/week4/2025-06-29-GuessNumber.java
package week4;
import java.util.*;

/*
 * Week 4: Binary Search
 * https://leetcode.com/problems/guess-number-higher-or-lower/
 */
class Solution {
    public static int guessNumber(int n) {
        int left = 1, right = n;

        while (left <= right) {
            int mid = left + (right - left) / 2;
            int result = guess(mid);        // pre-defined API

            if (result == 0) {
                return mid;
            } else if (result < 0) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        return -1;
    }

    public static void main(String[] args) {
        int[] input = {-1,0,3,5,9,12};
        guessNumber(10);
    }
}